src/providers/pkcs11/mod.rs

```diff
@@ -441,11 +446,34 @@ impl ProviderBuilder {
                 Error::new(ErrorKind::InvalidData, "error initializing PKCS 11 context")
             })?;
 
+
+        let slot_number = match self.slot_number {
+            Some(i) => {
+                let slot = Slot::try_from(i).or_else(|_| {
+                    Err(Error::new(
+                            ErrorKind::InvalidData,
+                            "cannot convert slot value",
+                    ))
+                })?;
+                slot
+            },
+            None => {
+                let slots = backend
+                    .get_slots_with_initialized_token().ok()
+                    .ok_or_else(|| Error::new(ErrorKind::InvalidData, "failed retreiving a valid slot with a initialized token"))?;
```

Comment:
You can use `map_err` here in order to keep the information on the error that happened:
```rust
                let slots = backend
                    .get_slots_with_initialized_token()
                    .map_err(|e| {
                        format_error!("Failed retrieving a valid slot with an initialized token", e);
                        Error::new(ErrorKind::InvalidData, "failed retrieving a valid slot with an initialized token")
                    })?;
```